nntrainer/tensor/int4_tensor.cpp

```diff
@@ -0,0 +1,504 @@
+    initialize();
+  }
+}
+
+void Int4QTensor::deallocate() {
+  data = nullptr;
+  offset = 0;
+}
+
+void *Int4QTensor::getData() const {
+  if (!data)
+    return nullptr;
+
+  data->validate();
+  return data->getAddr<int8_t>() + offset;
+}
+
+void *Int4QTensor::getData(size_t idx) const {
+  if (!data)
+    return nullptr;
+
+  data->validate();
+  return data->getAddr<int8_t>() + offset + (idx / 2);
+}
+
+void *Int4QTensor::getScale() const {
+  if (!data)
+    return nullptr;
+
+  data->validate();
+  return ((int8_t *)getData()) + (size() + 1) / 2;
+}
+
+void *Int4QTensor::getScale(size_t idx) const {
+  NNTR_THROW_IF(idx > scale_size(), std::invalid_argument)
+    << "Tensor::getScale() index is not valid";
+
+  if (!data)
+    return nullptr;
+
+  data->validate();
+  return ((float *)getScale()) + idx;
+}
+
+void *Int4QTensor::getAddress(unsigned int i) {
+  size_t index = getIndex(batch(), channel(), height(), width());
+  if (i > index) {
+    return nullptr;
+  }
+  return &((int8_t *)getData())[i / 2];
+}
+
+const void *Int4QTensor::getAddress(unsigned int i) const {
+  size_t index = getIndex(batch(), channel(), height(), width());
+  if (i > index) {
+    return nullptr;
+  }
+  return &((int8_t *)getData())[i / 2];
+}
+
+const int8_t Int4QTensor::getValue(unsigned int i) const {
+  int8_t value = ((int8_t *)getData())[i / 2];
+  return (i % 2 == 0) ? value >> 4 : ((int8_t)(value << 4) >> 4);
+}
+
+int8_t Int4QTensor::getValue(unsigned int i) {
+  int8_t value = ((int8_t *)getData())[i / 2];
+  return (i % 2 == 0) ? value >> 4 : ((int8_t)(value << 4) >> 4);
+}
+
+const int8_t Int4QTensor::getValue(unsigned int b, unsigned int c,
+                                   unsigned int h, unsigned int w) const {
+  return getValue(getIndex(b, c, h, w));
+}
+
+int8_t Int4QTensor::getValue(unsigned int b, unsigned int c, unsigned int h,
+                             unsigned int w) {
+  return getValue(getIndex(b, c, h, w));
+}
+
+/// @todo this func should be template function
+void Int4QTensor::setValue(float value) {
+  NNTR_THROW_IF(value < -8 || value > 7, std::out_of_range)
+    << "Value must be in range [-8, 7]. Input value: " << value;
+
+  int8_t val = value;
+  int8_t *data = (int8_t *)getData();
+  std::fill(data, data + (size() + 1) / 2, (val << 4) | (val & 0x0f));
+}
+
+/// @todo this func should be template function
+void Int4QTensor::addValue(unsigned int b, unsigned int c, unsigned int h,
+                           unsigned int w, float value, float beta) {
+  auto const &idx = getIndex(b, c, h, w);
+  float output = getValue(idx);
+  output *= beta;
+  output += value;
+
+  // if result value is out of range, clamp to max/min value
+  int8_t val = std::trunc(std::clamp((int)output, -8, 7));
+
+  // encode result value to int8 data
+  ((int8_t *)getData())[idx / 2] =
+    (idx % 2 == 0) ? (val << 4) | (((int8_t *)getData())[idx / 2] & 0x0f)
+                   : (((int8_t *)getData())[idx / 2] << 4) | (val & 0x0f);
```

Comment:
As I understood, isn't the computation should be 
```suggestion
                   : (((int8_t *)getData())[idx / 2] & 0xf0) | (val & 0x0f);
```
I'm quite confused with it. Please let me know If I'm wrong :)

Reply:
you're right! thanks for pointing it out :)